libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -2980,6 +2953,80 @@ string YulUtilFunctions::conversionFunction(Type const& _from, Type const& _to)
 	});
 }
 
+string YulUtilFunctions::arrayConversionFunction(ArrayType const& _from, ArrayType const& _to)
+{
+	string functionName =
+		"convert_array_" +
+		_from.identifier() +
+		"_to_" +
+		_to.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		Whiskers templ(R"(
+			function <functionName>(value<?fromCalldataDynamic>, length</fromCalldataDynamic>) -> converted {
+				<?sameTypes>
+					converted := value
+				<!sameTypes>
+					<?toMemory>
+						// Copy the array to a free position in memory, unless it is already in memory.
+						<?fromMemory>
+							converted := value
+						</fromMemory>
+						<?fromStorage>
+							converted := <arrayStorageToMem>(value)
+						</fromStorage>
+						<?fromCalldata>
+							converted := <allocateMemoryArray>(length)
+							<copyToMemory>(value, add(converted, 0x20), length)
+						</fromCalldata>
+					</toMemory>
+
+					<?toStorage>
+						converted := value
+					</toStorage>
+				</sameTypes>
+			}
+		)");
+		templ("functionName", functionName);
+		templ("sameTypes", _from == _to);
+		templ("toMemory", _to.dataStoredIn(DataLocation::Memory));
+		templ("toStorage", _to.dataStoredIn(DataLocation::Storage));
+		templ("fromMemory", _from.dataStoredIn(DataLocation::Memory));
+		templ("fromStorage", _from.dataStoredIn(DataLocation::Storage));
+		templ("fromCalldata", _from.dataStoredIn(DataLocation::CallData));
+		templ("fromCalldataDynamic", _from.dataStoredIn(DataLocation::CallData) && _from.isDynamicallySized());
+
+		switch (_to.location())
+		{
+		case DataLocation::Storage:
+			// Other cases are done explicitly in LValue::storeValue, and only possible by assignment.
+			solAssert(
+				(_to.isPointer() || (_from.isByteArray() && _to.isByteArray())) &&
+				_from.location() == DataLocation::Storage,
+				"Invalid conversion to storage type."
+			);
+			break;
+		case DataLocation::Memory:
+			if (_from.location() == DataLocation::CallData)
+			{
+				solUnimplementedAssert(!_from.baseType()->isDynamicallyEncoded(), "");
+				solUnimplementedAssert(_from.isByteArray() && _to.isByteArray(), "");
+				solUnimplementedAssert(_to.isDynamicallySized(), "");
```

Comment:
These assertions came from the removed code in `conversionFunctionSpecial`. Why are they `calldata` specific?

Reply:
That case (removed code) is actually copy from calldata dynamic array which is encoded like `(offset, len)` and that's what makes it to go to `conversionFunctionSpecial`